L15-PHP-Interceptors.md

```diff
@@ -0,0 +1,237 @@
+Title
+----
+* Author(s): ddyihai
+* Status: Draft
+* Implemented in: PHP
+
+## Abstract
+
+This proposal introduces the interceptor for gRPC PHP.
+__
+## Background
+
+The PHP language implementation for gRPC does not currently have an
+interceptor interface, whereas other languages like Go and Java do.
+Application concerns and extensibility are currently closed in the PHP
+implementation, making it difficult to extend the libraries for various
+application needs.
+
+
+### Related Proposals
+
+* [Python Client and Server Interceptors](https://github.com/grpc/proposal/pull/39)
+* [Ruby Client and Server Interceptors](https://github.com/grpc/proposal/pull/34)
+* [C# Client and Server Interceptors](https://github.com/grpc/proposal/pull/38)
+
+## Proposal
+
+Add client interceptors to the PHP gRPC implementation. Similar to the
+implementation in python, the channel is wrapper around an intercepted channel
+and the stub can use it the same way as the normal `Grpc\Channel`. RPCs dispatched 
+from such stubs then go through the interceptor chain for processing before they
+hit the underlying gRPC channel. 
+
+### Client Interceptors
+
+PHP provides a function `Grpc\Interceptor::intercept($channel, $interceptor)`
+to create a chain of interceptor, the API looks like below:
+ 
+```php
+Grpc\Interceptor::intercept(	// return intercepted channel
+	Channel				// underlying ‘Grpc\Channel’ or ‘Grpc\InterceptorChannel’
```

Comment:
I am hesitant in using the identifier `InterceptorChannel` anywhere in the proposal (using "intercepted channel" phrase is better as it is more general), except for clarification purposes, as I strongly wish that would be an implementation detail and not something this proposal actually proposes.